rasa/core/featurizers/single_state_featurizer.py

```diff
@@ -172,137 +334,158 @@ def _create_features(
             )
         ]
 
-    @staticmethod
-    def _to_sparse_sentence_features(
-        sparse_sequence_features: List["Features"],
-    ) -> List["Features"]:
-        return [
-            Features(
-                scipy.sparse.coo_matrix(feature.features.sum(0)),
-                FEATURE_TYPE_SENTENCE,
-                feature.attribute,
-                feature.origin,
-            )
-            for feature in sparse_sequence_features
-        ]
+    def _custom_encoding(self, sub_state: SubState, attribute: Text) -> Dict[Text, int]:
+        """Derives an encoding to be used by `_featurize_attribute_via_custom_encoding`.
 
-    def _get_features_from_parsed_message(
-        self, parsed_message: Optional[Message], attributes: Set[Text]
-    ) -> Dict[Text, List["Features"]]:
-        if parsed_message is None:
-            return {}
+        # NOTE: this was `_state_features_for_attribute`
 
-        output = defaultdict(list)
-        for attribute in attributes:
-            all_features = parsed_message.get_sparse_features(
-                attribute
-            ) + parsed_message.get_dense_features(attribute)
+        """
+        # FIXME: Create dataclasses for states, substates, messages,...
```

Comment:
I agree

Reply:
We should maybe start a notion page for those ... to keep links to places where those are used and in which way to figure out how the interfaces should look like?